[Resubmit for PR #1881] New option for SLUCM to use global distributed urban aerodynamic parameters

cenlinhe opened the pull request:
This is a re-submitted PR for https://github.com/wrf-model/WRF/pull/1881 to connect with latest Noah-MP GitHub commit and branch for v4.6 release. (Below is copied from https://github.com/wrf-model/WRF/pull/1881). For all review comments, please refer back to the original PR: https://github.com/wrf-model/WRF/pull/1881

New option for SLUCM to consider global distributed urban parameters

TYPE: new feature

KEYWORDS: SLUCM, urban parameters, anthropogenic heat

SOURCE: Do Ngoc Khanh (Tokyo Institute of Technology)

DESCRIPTION OF CHANGES:
This PR adds a new feature to WRF SLUCM by allowing consideration of spatially varying global distributed urban parameters and spatially hourly monthly varying anthropogenic heat.

LIST OF MODIFIED FILES:
M Registry/Registry.EM_COMMON
M Registry/registry.dimspec
M dyn_em/module_first_rk_step_part1.F
M dyn_em/module_initialize_real.F
M phys/module_pbl_driver.F
M phys/module_physics_init.F
M phys/module_sf_clm.F
M phys/module_sf_noahdrv.F
M phys/module_sf_urban.F
M phys/module_surface_driver.F
M phys/noahmp
M share/output_wrf.F

TESTS CONDUCTED:

The modification has been tested and used in previous publications.
Initial development: Varquez, A. C. G., Nakayoshi, M., & Kanda, M. (2015). The effects of highly detailed urban roughness parameters on a sea-breeze numerical simulation. Boundary-layer meteorology, 154, 449-469. https://doi.org/10.1007/s10546-014-9985-4
Global extension: Khanh, D. N., Varquez, A. C., & Kanda, M. (2023). Impact of urbanization on exposure to extreme warming in megacities. Heliyon, 9, e15511. https://doi.org/10.1016/j.heliyon.2023.e15511
Are the Jenkins tests all passing?
RELEASE NOTE: This modification add a two options (use_distributed_aerodynamics and distributed_ahe_opt) to WRF SLUCM (sf_urban_physics = 1) so that spatially varying urban morphological parameters (building height, plan area index, frontal area index, roughness length for momentum, and displacement height) can be considered.

cenlinhe commented:
For all review comments, please refer back to the original PR: https://github.com/wrf-model/WRF/pull/1881

weiwangncar commented:
@cenlinhe The compilation for the model has failed. I'm sending you one of the output files.

cenlinhe commented:
There was a minor bug and I have fixed it. I just tested it on Derecho and it was successfully compiled.

cenlinhe commented:
@epn09 I resubmitted your PR with updates from Noah-MP GitHub. Please take a look to see if I missed anything.

weiwangncar commented:
The regression test results:

    Test Type              | Expected  | Received |  Failed
    = = = = = = = = = = = = = = = = = = = = = = = =  = = = =
    Number of Tests        : 23           24
    Number of Builds       : 60           57
    Number of Simulations  : 158           150        0
    Number of Comparisons  : 95           86        0

    Failed Simulations are: 
    None
    Which comparisons are not bit-for-bit: 
    None

epn09 commented:
@cenlinhe Thank you very much. I think the change was correctly reflected in this resubmitted PR.

dudhia reviewed (commented):
Not sure why it is called distributed aerodynamics. Add the info to the README.namelist.

cenlinhe commented:
This PR needs to be merged after PR#1971 because of the update from Noah-MP submodule

cenlinhe commented:
> Not sure why it is called distributed aerodynamics. Add the info to the README.namelist.

I will add info to the README.namelist today.

weiwangncar commented:
@epn09 Can you provide input file (wrfinput and wrfbdy), for one domain for us to try? Can you also describe the data you use to generate the input. Did you create a binary data for input to geogrid program? Is the data global? If this is the case, you should have a modified GEOGRID.TBL file, and you can make a PR to update that file, and provide the binary data directly to us.

epn09 commented:
> Not sure why it is called distributed aerodynamics. Add the info to the README.namelist.

@dudhia We also can't traced back the origin of this name. If you have suggestions on naming. Please tell us.

epn09 commented:
@weiwangncar Please download a test run for Tokyo from our lab server here: https://urbanclimate.tse.ens.titech.ac.jp/database/for_wrf_pr/

The binary files for WPS are uploaded here: https://urbanclimate.tse.ens.titech.ac.jp/database/AHE/AH4GUC/geogrid/
I will make a PR to WPS with modified GEOGRID.TBL soon.

weiwangncar commented:
@epn09 Is this a global dataset? What is its resolution?

weiwangncar commented on Registry/Registry.EM_COMMON:
```diff
@@ -807,13 +808,18 @@ state    real   DZR             l        em      -         Z     r        "DZR"
+state    real   Z0_URB2D       ij       misc      1        -     i01r     "Z0_URB2D"   "Roughness length for momentum"  "m"
+state    real   LF_URB2D_S     ij       misc      1        -     i01r     "LF_URB2D_S" "Frontal area index (no wind directional dependency)" ""
+# AHE with month and hour dimension flattened to one dimension, Jan = (0:23), Feb = (24:47)
+state    real   AHE            i{m_hr}j  misc       1        -     i01r   "AHE"        "Anthropogenic heat emission" "W m-2"
```
@cenlinhe You should consider packaging these new arrays to reduce memory if the option is not used. Once the variable is part of the registry, I don't think you can use optional variable any more. They are in the code, present. If people don't have it in the metgrid input files, and the package sets the arrays to a dimension of 1, and the code should have conditionals to prevent accessing/using these arrays.

dudhia replied:
Agreed. Looks safe to package it and it is used with options 1 and 2 of this namelist variable.

weiwangncar commented on phys/module_pbl_driver.F:
```diff
@@ -2204,6 +2214,25 @@ SUBROUTINE pbl_driver(                                          &
               ,ITS=its,ITE=ite,JTS=jts,JTE=jte,KTS=kts,KTE=kte)
           ENDIF
 
+     IF (PRESENT(ahe)) THEN
```
@cenlinhe This is one place I am referring to. Use namelist to control the code, not the present of an array.

epn09 commented:
> Is this a global dataset? What is its resolution?

@weiwangncar The datasets are global at 30-arcsecond spatial resolution.

weiwangncar commented:
> > Is this a global dataset? What is its resolution?
> 
> @weiwangncar The datasets are global at 30-arcsecond spatial resolution.

Thank you for the information.

epn09 commented:
@weiwangncar @cenlinhe I've just made a pull request to WPS https://github.com/wrf-model/WPS/pull/244.

dudhia commented on dyn_em/module_initialize_real.F:
```diff
@@ -3089,7 +3104,17 @@ SUBROUTINE init_domain_rk ( grid &
+            DO i = its , MIN(ide-1,ite)
+              IF (grid%ivgtyp(i, j) == model_config_rec%isurban(grid%id)) THEN
+                grid%frc_urb2d(i, j) = MAX(0.1, MIN(0.9, 1 - grid%shdavg(i, j) / 100.))
+              END IF
```
urban fraction set to annual mean vegetation fraction?

cenlinhe replied:
I asked the same question to the developer and he said this treatment seems most reasonable for their new scheme.

dudhia replied:
I would be concerned about urban areas in dry/desert regions. Why not trust frc_urb2d more?

epn09 replied:
@dudhia
* Is FRC_URB2D available globally? I am sorry for my lack of understanding but I've never used that static data before.
* May be we can add a switch to allow users to pick between our way of estimating FRC_URB2D and the static FRC_URB2D input file.

dudhia replied:
we can think of it later

smileMchen commented:
A few issues here:
(1)  Both WPS-SLUCM and WRF-SLUCM can be successfully compiled in derecho.
(2) All the data for SLUCM have been tested and saved in derecho (/glade/derecho/chenming/SLUCMDATA/). We will put these data to wrfhelp/WPS_GEOG soon. 
(3) The following new options have been tested and they all work fine
      use_distributed_aerodynamics = true
      distributed_ahe_opt =0
      distributed_ahe_opt =1
      distributed_ahe_opt =2
(4) Note that this PR-related Modis data doesn't include 'lake', and hence the num_land_cat should be 20.
      I would suggest we use our original MODIS data (21 categories) instead of this new Modis data unless this new data has certain advantages.

epn09 commented:
@smileMchen 

> (4) Note that this PR-related Modis data doesn't include 'lake', and hence the num_land_cat should be 20.
I would suggest we use our original MODIS data (21 categories) instead of this new Modis data unless this new data has certain advantages.

As written in the PR in WPS (https://github.com/wrf-model/WPS/pull/244), the modis data in our simulation was the modis_landuse_20class_15s (without lakes) where all grids with population density > 1000/km^2 set to urban. The reason is because we found that urban areas in modis_landuse_20class_15s were too small. 

I'm not sure if this can be considered an advantage. Maybe it's more like a quick fix.
Of course if with-lakes data is preferable, we can do the same adjustment to modis_landuse_20class_15s_with_lakes

dudhia commented:
> > Not sure why it is called distributed aerodynamics. Add the info to the README.namelist.
> 
> @dudhia We also can't trace back the origin of this name. If you have suggestions on naming. Please tell us.

I would suggest a name like ucm_distributed_drag

smileMchen commented:
> @smileMchen
> 
> > (4) Note that this PR-related Modis data doesn't include 'lake', and hence the num_land_cat should be 20.
> > I would suggest we use our original MODIS data (21 categories) instead of this new Modis data unless this new data has certain advantages.
> 
_> As written in the PR in WPS ([wrf-model/WPS#244](https://github.com/wrf-model/WPS/pull/244)), the modis data in our simulation was the modis_landuse_20class_15s (without lakes) where all grids with population density > 1000/km^2 set to urban. The reason is because we found that urban areas in modis_landuse_20class_15s were too small.
> 
> I'm not sure if this can be considered an advantage. Maybe it's more like a quick fix. Of course if with-lakes data is preferable, we can do the same adjustment to modis_landuse_20class_15s_with_lakes_

@epn09 
I did see larger areas of urban landuse type  using your Modis data. Since this PR is focused on urban physics, I suppose users will see more obvious impacts using the new Modis data. For this reason, I would prefer to keep this data. Let's see what others think. We can make a decision later.

cenlinhe commented:
> > @smileMchen
> > > (4) Note that this PR-related Modis data doesn't include 'lake', and hence the num_land_cat should be 20.
> > > I would suggest we use our original MODIS data (21 categories) instead of this new Modis data unless this new data has certain advantages.
> 
> _> As written in the PR in WPS ([wrf-model/WPS#244](https://github.com/wrf-model/WPS/pull/244)), the modis data in our simulation was the modis_landuse_20class_15s (without lakes) where all grids with population density > 1000/km^2 set to urban. The reason is because we found that urban areas in modis_landuse_20class_15s were too small.
> 
> > I'm not sure if this can be considered an advantage. Maybe it's more like a quick fix. Of course if with-lakes data is preferable, we can do the same adjustment to modis_landuse_20class_15s_with_lakes_
> 
> @epn09 I did see larger areas of urban landuse type using your Modis data. Since this PR is focused on urban physics, I suppose users will see more obvious impacts using the new Modis data. For this reason, I would prefer to keep this data. Let's see what others think. We can make a decision later.

I am fine with keeping this data but needs to have a clear notes in WPS and/or WRF readme files and user guide to remind users to only use this new dataset when they activate the new urban capabilities implemented in this WRF PR.

dudhia commented:
For those of us not following this closely, is the new urban dataset required for this option to work because it has extra necessary inputs?

cenlinhe commented:
> For those of us not following this closely, is the new urban dataset required for this option to work because it has extra necessary inputs?

Yes, it requires the new urban dataset because it has extra necessary inputs.

smileMchen commented:
> > > @smileMchen
> > > > (4) Note that this PR-related Modis data doesn't include 'lake', and hence the num_land_cat should be 20.
> > > > I would suggest we use our original MODIS data (21 categories) instead of this new Modis data unless this new data has certain advantages.
> > 
> > 
> > _> As written in the PR in WPS ([wrf-model/WPS#244](https://github.com/wrf-model/WPS/pull/244)), the modis data in our simulation was the modis_landuse_20class_15s (without lakes) where all grids with population density > 1000/km^2 set to urban. The reason is because we found that urban areas in modis_landuse_20class_15s were too small.
> > > I'm not sure if this can be considered an advantage. Maybe it's more like a quick fix. Of course if with-lakes data is preferable, we can do the same adjustment to modis_landuse_20class_15s_with_lakes_
> > 
> > 
> > @epn09 I did see larger areas of urban landuse type using your Modis data. Since this PR is focused on urban physics, I suppose users will see more obvious impacts using the new Modis data. For this reason, I would prefer to keep this data. Let's see what others think. We can make a decision later.
> 
> I am fine with keeping this data but needs to have a clear notes in WPS and/or WRF readme files and user guide to remind users to only use this new dataset when they activate the new urban capabilities implemented in this WRF PR.

@cenlinhe 
I agree.

cenlinhe commented:
In my latest commit, as suggested by Wei and Jimy, I have:
1. packaged the new variables in registry
2. renamed the namelist option from urban_distributed_aerodynamics to slucm_distributed_drag
3. replaced the "if present(ahe)" by "if (distributed_ahe_opt > 0)".

weiwangncar commented:
@cenlinhe Thanks for renaming the namelist and package the arrays. Now the PR has some conflict with two files: phys/module_pbl_driver.F and phys/noahmp/. Can you help resolve the conflicts?

cenlinhe commented:
> @cenlinhe Thanks for renaming the namelist and package the arrays. Now the PR has some conflict with two files: phys/module_pbl_driver.F and phys/noahmp/. Can you help resolve the conflicts?

I cannot see the conflicting parts. Also for noahmp, I am not sure why but the commit in Noah-MP side should be good which should not lead to any conflicts. The pbl driver part probably includes some newly merged update in other parts inside this fortran file. I will see what I can do.

weiwangncar commented:
@cenlinhe Thanks for resolving the conflicts. The last change passed regression tests. We should be able to approve this PR now. 
@dudhia Can you review this PR?

weiwangncar commented on phys/module_pbl_driver.F:
```diff
@@ -2260,6 +2269,25 @@ SUBROUTINE pbl_driver(                                          &
+         END DO
+       END IF
+     END IF
+
```
@epn09 @cenlinhe Why is HFX added at the end of PBL driver? What is the added HFX for?

cenlinhe replied:
this is for anthropogenic heat flux (i.e., sensible heat flux) released from surface human activities to the atmosphere.

weiwangncar replied:
@cenlinhe After PBL call, how do you expect the flux to be affecting the atmosphere?

cenlinhe replied:
I think @epn09 originally wants to let this flux affect atmosphere for the next time step. Another technical consideration is that @epn09 wants to add this flux to the temperature tendency (ahe_option=2) which is an output of PBL scheme, so this part needs to be added at the end of PBL call. This is my understanding if I remember his earlier explanations.

epn09 replied:
@weiwangncar @cenlinhe 
* I think we added anthropogenic heat (AH) to the end of PBL so that it will not be overwritten by any other scheme.
* Already discussed before but AH is added in PBL instead of UCM because AH can present in non-urban grids.
* There are two options for AH:
   1. Add to the temperature tendency at the first layer of the atmosphere
   2. Add directly to the surface sensible heat flux.
* So far, we found that the first option often produce more natural (?) result but we'd like to expose both options to WRF users.

epn09 replied:
@weiwangncar I see. So where would be the right place to add AH to HFX?

weiwangncar replied:
@epn09 Perhaps you could do this in module_pbl_driver.F, but before any PBL physics is called.

dudhia replied:
Some PBL schemes may set RTHBLTEN to zero before adding to it. It needs to be checked.

epn09 commented on Registry/Registry.EM_COMMON:
```diff
@@ -3127,7 +3135,7 @@ package   temfsfcscheme      sf_sfclay_physics==10       -             state:wm_
 package   sfclayscheme       sf_sfclay_physics==91       -             -
 
-package   noahucmscheme  sf_urban_physics==1         -             state:trb_urb4d,tw1_urb4d,tw2_urb4d,tgb_urb4d,sfw1_urb3d,sfw2_urb3d,sfr_urb3d,sfg_urb3d,a_u_bep,a_v_bep,a_t_bep,a_q_bep,a_e_bep,b_u_bep,b_v_bep,b_t_bep,b_q_bep,b_e_bep,dlg_bep,dl_u_bep,sf_bep,vl_bep,mh_urb2d,stdh_urb2d,lf_urb2d,lp_urb2d,hgt_urb2d,lb_urb2d,tgr_urb2d,cmcr_urb2d,drelr_urb2d,drelb_urb2d,drelg_urb2d,flxhumr_urb2d,flxhu ...
+package   noahucmscheme  sf_urban_physics==1         -             state:trb_urb4d,tw1_urb4d,tw2_urb4d,tgb_urb4d,sfw1_urb3d,sfw2_urb3d,sfr_urb3d,sfg_urb3d,a_u_bep,a_v_bep,a_t_bep,a_q_bep,a_e_bep,b_u_bep,b_v_bep,b_t_bep,b_q_bep,b_e_bep,dlg_bep,dl_u_bep,sf_bep,vl_bep,mh_urb2d,stdh_urb2d,lf_urb2d,lp_urb2d,hgt_urb2d,lb_urb2d,tgr_urb2d,cmcr_urb2d,drelr_urb2d,drelb_urb2d,drelg_urb2d,flxhumr_urb2d,flxhu ...
```
@cenlinhe I don't understand `package` in Registry very well but should we add here something like
```
package ahescheme distributed_ahe_opt==1 ahe
package ahescheme distributed_ahe_opt==2 ahe
```
so that those who don't use AH won't need to allocate a big array?

cenlinhe replied:
I think it is fine for now to package it to SLUCM scheme. Later, we can create a new package if needed. Currently, there is limited time to finalize the code for the upcoming WRF release, so I do not suggest any major changes unless it is necessary.

epn09 replied:
@cenlinhe Thank you. I understand.
